cdap-docs/developers-manual/source/building-blocks/mapreduce-programs.rst

```diff
@@ -161,12 +182,147 @@ declaration and (2) an injection:
+
+       private MapReduceTaskContext taskContext;
+
+       @Override
+       public void initialize(MapReduceTaskContext mapReduceTaskContext) throws Exception {
+         taskContext = mapReduceTaskContext;
+       }
+
+       @Override
+       public void destroy() {
+       }
+
+       @Override
+       public void map(byte[] key, Purchase purchase, Context context)
+           throws IOException, InterruptedException {
+         // join with catalog by product ID
+         String catalogName = determineCatalogName(purchase.getProductCategory());
+         ProductCatalog catalog = taskContext.getDataset(catalogName);
+         Product product = catalog.read(purchase.getProductId());
+         ...
+       }
+
+       private String determineCatalogName(String productCategory) {
+         ...
+       }
+
+See also the section on :ref:`Using Datasets in Programs <datasets-in-programs>`.
 
 .. rubric: Datasets as MapReduce Input or Output
 
-Additionally, a MapReduce program can interact with a dataset by using it as an input or an
-output, as described in :ref:`datasets-mapreduce-programs`.
+A MapReduce program can interact with a dataset by using it as 
+:ref:`an input <mapreduce-datasets-input>` or :ref:`an output <mapreduce-datasets-output>`.
+The dataset needs to implement specific interfaces to support this, as described in the
+following sections.
+
+.. _mapreduce-datasets-input:
+
+.. rubric:: A Dataset as the Input Source of a MapReduce Program
+
+When you run a MapReduce program, you can configure it to read its input from a dataset. The
+source dataset must implement the ``BatchReadable`` interface, which requires two methods::
+
+  public interface BatchReadable<KEY, VALUE> {
+    List<Split> getSplits();
+    SplitReader<KEY, VALUE> createSplitReader(Split split);
+  }
+
+These two methods complement each other: ``getSplits()`` must return all splits of the dataset
+that the MapReduce program will read; ``createSplitReader()`` is then called in every Mapper to
+read one of the splits. Note that the ``KEY`` and ``VALUE`` type parameters of the split reader
+must match the input key and value type parameters of the Mapper.
+
+Because ``getSplits()`` has no arguments, it will typically create splits that cover the
+entire dataset. If you want to use a custom selection of the input data, define another
+method in your dataset with additional parameters and explicitly set the input in the
+``beforeSubmit()`` method.
+
+For example, the system dataset ``KeyValueTable`` implements ``BatchReadable<byte[], byte[]>``
+with an extra method that allows specification of the number of splits and a range of keys::
+
+  public class KeyValueTable extends AbstractDataset
+                             implements BatchReadable<byte[], byte[]> {
+    ...
+    public List<Split> getSplits(int numSplits, byte[] start, byte[] stop);
+  }
+
+To read a range of keys and give a hint that you want 16 splits, write::
+
+  @Override
```

Comment:
I think this is a concern with the existing documentation, but I think the `@Override` should be moved down to right above the `public void beforeSubmit(MapReduceContext context) throws Exception {`.